apollo-backend/src/main/java/io/logz/apollo/kubernetes/KubernetesMonitor.java

```diff
@@ -104,7 +104,12 @@ public void monitor() {
 
                 switch (deployment.getStatus()) {
                     case PENDING:
-                        returnedDeployment = kubernetesHandler.startDeployment(deployment);
+                        if (isDeployedEnvironmentConcurrencyLimitPermitsDeployment(deployment)) {
```

Comment:
This code does not seem thread safe. That means that if more than one thread checks if it is allowed to deploy, two might get the "ok" and we'll end up with more than the concurrency limit of running deployments. Is that a possible use case here (multi threaded)?

Reply:
Its a Singleton by design, Apollo is not thread safe and not multi-process safe. We plan on adding Apollo roles and consensus for cluster deployment of Apollo, but not yet